### PR TITLE
Add the ability to merge multiple morph-with-spines files into one

### LINKED ISSUE #37: New feature: merge multiple morph-with-spines files into one
We need a new API function that takes multiple morph-with-spines files and merges them into a single (new) file, with, optionally, group key renaming (either neuron names or spines library names).

### PR DESCRIPTION
**Pull request by jplanasc**
Fixes #37 

# Description
Add `merge_morphologies_with_spines` API function to merge multiple morphology-with-spines HDF5 files into a single output file, with optional renaming of neuron keys and spines library names.

## Features
- Merge any number of morph-with-spines files into one output
- Rename neuron keys and/or spines library names via a `rename_map` keyed by `(source_file_path, original_name)`
- `spine_morphology` entry in the spines tables is updated on the fly during copy
- Supports duplicate input morph/spine names across source files as long as they're renamed to unique destinations
- `include_meshes` parameter to optionally exclude soma and spine mesh data in the output file
- Validation: fail-fast, before any output is written:
  - Source files must be valid morph-with-spines format
  - All `spine_morphology` values must reference existing groups in input file's /spines/skeletons group
  - Destination names must be unique (both morphology keys and spines library names, checked independently)

## Testing
- Input validation (empty files, missing groups, duplicates)
- Merge without renaming (single file, multiple files, multi-neuron files, data preservation)
- Merge with renaming (neuron rename, shared library rename, overlap detection, longer names)
- include_meshes test
- Edge cases (missing soma meshes, missing referenced groups, spines shared library)

## Documentation
- Updated README with examples

### CODE REVIEW

File: src/morph_spines/utils/morph_spine_merger.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+def _validate_sources(
+    source_files: list[Path],
+    rename_map: dict[tuple[Path, str], str],
+) -> None:
+    """Validate source file format and check for destination name collisions."""
+    morph_dest_names: list[str] = []
+    spine_dest_names: list[str] = []
+
+    for src_path in source_files:
+        with h5py.File(src_path, "r") as h5:
```
Comment (MWolfR) [inline src/morph_spines/utils/morph_spine_merger.py:109]: May I suggest to move lines 94 to 108 into a separate function that takes the path to a single file as input, checks its consistency and then returns
`(list(h5[GRP_MORPH].keys()), list(skeletons_grp.keys()))`
or raises an Exception if the file is invalid.

That function could then serve as a standalone validation and provides a quick overview of the contents of a morphology-with-spines. That would be very useful in many contexts.

Then _validate_sources would call that function and the rest would remain unchanged. Simple cut-and-paste.

File: src/morph_spines/utils/morph_spine_merger.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+        raise ValueError("source_files must not be empty")
+    if output_path.exists():
+        raise FileExistsError(f"Output file already exists: {output_path}")
+
+    rename_map = rename_map or {}
+
+    _validate_sources(source_files, rename_map)
+
+    # Pre-compute per-file name maps
+    file_name_maps: dict[Path, dict[str, str]] = {}
```
Comment (MWolfR) [inline src/morph_spines/utils/morph_spine_merger.py:82]: **Very optional:**
The function signature clearly specifies that the file name inputs must be specified as `Path`. But let's face it: Some people will provide `str` as inputs.

I tested that case and it actually works without trouble! The only problem arises when you use `Path` in `source_files`, but the keys of `rename_map` use `(str, str)`. Then `rename_map` gets silently ignored!

This could be solved by explicitly turning the paths of both arguments into `Path` objects. 
I realize that this is an edge case that requires an almost intentionally stupid user, so if you say it's out of scope, feel free to ignore.

File: src/morph_spines/utils/morph_spine_merger.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+    _check_duplicates(morph_dest_names, "morphology")
+    _check_duplicates(spine_dest_names, "spines library")
+
+
+def _check_duplicates(names: list[str], label: str) -> None:
+    """Raise ValueError if any name appears more than once."""
+    seen: set[str] = set()
+    for name in names:
+        if name in seen:
+            raise ValueError(f"Duplicate {label} destination name: '{name}'")
```
Comment (MWolfR) [inline src/morph_spines/utils/morph_spine_merger.py:161]: Wouldn't 
`if len(names) != len(set(names)):`
do this simpler?

File: src/morph_spines/utils/morph_spine_merger.py [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+        new_name = name_map.get(old_name, old_name)
+        if new_name != old_name:
+            spine_morph_values[spine_morph_values == old_name] = new_name
+            needs_update = True
+
+    dt = h5py.string_dtype(encoding="utf-8")
+    for ds_name in src_grp:
+        if isinstance(src_grp[ds_name], h5py.Group):
+            # Metadata subgroup
+            h5_in.copy(src_grp[ds_name], dst_grp, name=ds_name)
```
Comment (MWolfR) [inline src/morph_spines/utils/morph_spine_merger.py]: That is the same as line 215..?
Why the distinction?

**Review by MWolfR — CHANGES_REQUESTED** [review]
Individual comments / ideas.

Plus: 
Maybe if `rename_map` contains an entry that is never used a warning should be raised? Alternative: an optional `strict: bool = False`, which if set to True raises an Exception if `rename_map` contains unused entries.

File: src/morph_spines/utils/morph_spine_merger.py (hunk 2 above)
Comment (jplanasc) [reply src/morph_spines/utils/morph_spine_merger.py:82]: Very good catch! Fixed.

File: src/morph_spines/utils/morph_spine_merger.py (hunk 1 above)
Comment (jplanasc) [reply src/morph_spines/utils/morph_spine_merger.py:109]: I'm working on a separate PR that will specifically check file integrity, so I didn't invest much in this part here, as it will be replaced by the following PR (to be created, check https://github.com/openbraininstitute/morph-spines/issues/39 ).

I propose we revisit this once I finish the file validator PR.

File: src/morph_spines/utils/morph_spine_merger.py (hunk 3 above)
Comment (jplanasc) [reply src/morph_spines/utils/morph_spine_merger.py:161]: Yes, but the advantage of the current code is that it tells you _which_ duplicates it found.

File: src/morph_spines/utils/morph_spine_merger.py (hunk 4 above)
Comment (jplanasc) [reply src/morph_spines/utils/morph_spine_merger.py]: Indeed, the branching can be simplified. Fixed.

File: src/morph_spines/utils/morph_spine_merger.py (hunk 3 above)
Comment (MWolfR) [reply src/morph_spines/utils/morph_spine_merger.py:161]: Good point :)

File: src/morph_spines/utils/morph_spine_merger.py (hunk 1 above)
Comment (MWolfR) [reply src/morph_spines/utils/morph_spine_merger.py:109]: Thumbs up to that.

**Comment by jplanasc** [comment]
> Maybe if rename_map contains an entry that is never used a warning should be raised?

Yes, makes sense. I added a warning.

**Review by MWolfR — APPROVED** [review]
